### PR TITLE
Not constructor arguments are allowed in static classes

### PR DESCRIPTION
**Pull request by edgarfgp**
Fixes one the multiple missing error reporting in https://github.com/dotnet/fsharp/issues/13147
 This PR adds: 
 - New compiler error when using constructor arguments on static classes `[<Sealed; AbstractClass>]`
 - The error is generic enough that we can reuse it for the rest of missing reporting

I will adding more error reporting in the future :)

### CODE REVIEW

**Comment by edgarfgp** [comment]
@T-Gro Any idea why the CI is failing ? . It seem to be related to my changes but Im failing to understand why :(

**Comment by T-Gro** [comment]
> @T-Gro Any idea why the CI is failing ? . It seem to be related to my changes but Im failing to understand why :(

Will have a look

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 1]
```diff
@@ -4047,7 +4051,13 @@ module TcDeclarations =
...
                 let membersIncludingAutoProps = 
                     cspec |> List.filter (fun memb -> 
-                      match memb with 
+                      match memb with
+                      | SynMemberDefn.ImplicitCtor(ctorArgs= SynSimplePats.SimplePats(pats, _)) when  not pats.IsEmpty && reportErrorOnStaticClass ->
+                            for pat in pats do
+                                 errorR(Error(FSComp.SR.chkErrorOnStaticClasses "constructor arguments are", pat.Range))
+                            false
+                      | SynMemberDefn.Member(SynBinding(valData = SynValData(Some memberFlags, _, _)), m)  when memberFlags.MemberKind = SynMemberKind.Constructor && reportErrorOnStaticClass ->
+                          errorR(Error(FSComp.SR.chkErrorOnStaticClasses "additional constructors are", m)); false
```
Comment (T-Gro) [inline src/Compiler/Checking/CheckDeclarations.fs:4060]: Bear in mind that all the text in FsComp.txt gets localized, whereas strings in compiled code are not.
I think it is OK to use parametrized templates, but then the embedded snippets should be as easy as possible, and typically not affect the form of the sentence even if different languages.

**Comment by T-Gro** [comment]
> @T-Gro Any idea why the CI is failing ? . It seem to be related to my changes but Im failing to understand why :(

Made a few pushes already:
- Repro test
- Make it work for langversion 7.0 (taking your new code path only when really needed), still fails for preview
- Now the issue is isolated, I think it is the call to TcAttributes  -> it is used "too early". 
In a mutually recursive type definition, type-checking the attribute might fail if the rest of mutually recursive definition has not been processed yet. This is exactly the case: **[<DebuggerTypeProxyAttribute(typedefof<{some type defined later...}<_>>)>]**

**Comment by T-Gro** [comment]
@edgarfgp Check the current code and commit log.
Possibilities I see:
- Do your custom defensive "TcAttributes", that will only search for known built-in ones and not eagerly typecheck, especially not typechecking arguments of attributes
- OR; move the entire logic to a later phase, where type-checked attributes already exist.

We can chat on Slack if you want to discuss this

**Comment by edgarfgp** [comment]
> @edgarfgp Check the current code and commit log. Possibilities I see:
> 
> * Do your custom defensive "TcAttributes", that will only search for known built-in ones and not eagerly typecheck, especially not typechecking arguments of attributes
> * OR; move the entire logic to a later phase, where type-checked attributes already exist.
> 
> We can chat on Slack if you want to discuss this

Thanks @T-Gro  :) .

**Comment by edgarfgp** [comment]
@T-Gro I found a later phase to do the attribute checking (Thanks for your help). Now that the CI is happy, I wanted to double-check if that error message and the ranges are clear enough. Any feedback is welcome :)

File: src/Compiler/FSComp.txt [hunk 2]
```diff
@@ -1665,4 +1665,5 @@ reprStateMachineInvalidForm,"The state machine has an unexpected form"
 3549,tcSynTypeOrInvalidInDeclaration,"SynType.Or is not permitted in this declaration"
 3550,chkDuplicatedMethodParameter,"Duplicate parameter. The parameter '%s' has been used more that once in this method."
 featureEscapeBracesInFormattableString,"Escapes curly braces before calling FormattableStringFactory.Create when interpolated string literal is typed as FormattableString"
-3551,chkErrorOnStaticClasses,"if a type uses both [<Sealed>] and [<AbstractClass>] it means it is static. No %s allowed"
\ No newline at end of file
+3551,buildDuplicateFile,"The source file '%s' (at position %d/%d) already appeared in the compilation list (at position %d/%d). Please verify that it is included only once in the project file."
+3552,chkErrorOnStaticClasses,"if a type uses both [<Sealed>] and [<AbstractClass>] it means it is static. No %s allowed"
```
Comment (T-Gro) [inline src/Compiler/FSComp.txt:1669]:
```suggestion
3552,chkErrorOnStaticClasses,"If a type uses both [<Sealed>] and [<AbstractClass>] attributes, it means it is static. %s is not allowed."
```

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 3]
```diff
@@ -1654,6 +1654,18 @@ module MutRecBindingChecking =
              mBinds 
         
         defnsEs, envMutRec
+        
+let private CheckStaticClassElements (synMembers: SynMemberDefn list) =
+    for mem in synMembers do
+        match mem with
+        | SynMemberDefn.ImplicitCtor(ctorArgs= SynSimplePats.SimplePats(pats= pats)) when  (not pats.IsEmpty) ->
+            for pat in pats do
+                errorR(Error(FSComp.SR.chkErrorOnStaticClasses "constructor arguments are", pat.Range))
```
Comment (T-Gro) [inline src/Compiler/Checking/CheckDeclarations.fs:1663]:
```suggestion
                errorR(Error(FSComp.SR.chkErrorOnStaticClasses "constructor with arguments", pat.Range))
```

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 4]
```diff
@@ -1654,6 +1654,18 @@ module MutRecBindingChecking =
...
+           
+        | SynMemberDefn.Member(SynBinding(valData = SynValData(memberFlags = Some memberFlags)), m)  when memberFlags.MemberKind = SynMemberKind.Constructor ->
+            errorR(Error(FSComp.SR.chkErrorOnStaticClasses "additional constructors are", m));
```
Comment (T-Gro) [inline src/Compiler/Checking/CheckDeclarations.fs:1666]:
```suggestion
            errorR(Error(FSComp.SR.chkErrorOnStaticClasses "additional constructor", m));
```

**Comment by T-Gro** [comment]
Left a suggestion.
I think we can leave certain nouns english and having rest of the sentence localized.

I would rather avoid putting verbs (are,is,..) into in and want to have them as part of the template, because different languages have different word order rules as well as negation rules. As long as it is the template, translator can change both the sentence as well as placement of the '%s' expression.

**Comment by vzarytovskii** [comment]
As discussed on slack, we should probably use different errors for each case, as oppose to having one interpolated (which won't go well with localizations).

File: src/Compiler/Checking/CheckDeclarations.fs (hunk 1 above)
Comment (edgarfgp) [reply src/Compiler/Checking/CheckDeclarations.fs]: I have updated to use 2 different error numbers so this is not longer the case :)

File: src/Compiler/FSComp.txt (hunk 2 above)
Comment (edgarfgp) [reply src/Compiler/FSComp.txt]: I have updated to use 2 different error numbers so this is not longer the case :)

File: src/Compiler/Checking/CheckDeclarations.fs (hunk 3 above)
Comment (edgarfgp) [reply src/Compiler/Checking/CheckDeclarations.fs]: I have updated to use 2 different error numbers so this is not longer the case :)

**Comment by edgarfgp** [comment]
@vzarytovskii Updated as discussed in Slack :)

**Review by T-Gro — APPROVED** [review]

**Comment by edgarfgp** [comment]
@0101   by Any chance I have another review and hopefully merge it?  So I can continue adding errors on static classes?

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 5]
```diff
@@ -1655,6 +1655,17 @@ module MutRecBindingChecking =
...
         defnsEs, envMutRec
 
+let private ReportErrorOnStaticClass (synMembers: SynMemberDefn list) =
+    for mem in synMembers do
+        match mem with
+        | SynMemberDefn.ImplicitCtor(ctorArgs= SynSimplePats.SimplePats(pats= pats)) when  (not pats.IsEmpty) ->
+            for pat in pats do
+                errorR(Error(FSComp.SR.chkConstructorWithArgumentsOnStaticClasses(), pat.Range))
+           
+        | SynMemberDefn.Member(SynBinding(valData = SynValData(memberFlags = Some memberFlags)), m)  when memberFlags.MemberKind = SynMemberKind.Constructor ->
```
Comment (0101) [inline src/Compiler/Checking/CheckDeclarations.fs:1665]:
```suggestion
        | SynMemberDefn.ImplicitCtor(ctorArgs = SynSimplePats.SimplePats(pats = pats)) when (not pats.IsEmpty) ->
            for pat in pats do
                errorR(Error(FSComp.SR.chkConstructorWithArgumentsOnStaticClasses(), pat.Range))
           
        | SynMemberDefn.Member(SynBinding(valData = SynValData(memberFlags = Some memberFlags)), m) when memberFlags.MemberKind = SynMemberKind.Constructor ->
```
Consistent whitespace.

File: src/Compiler/FSComp.txt [hunk 6]
```diff
@@ -1663,5 +1664,7 @@ reprStateMachineInvalidForm,"The state machine has an unexpected form"
...
 3549,tcSynTypeOrInvalidInDeclaration,"SynType.Or is not permitted in this declaration"
 3550,chkDuplicatedMethodParameter,"Duplicate parameter. The parameter '%s' has been used more that once in this method."
+featureEscapeBracesInFormattableString,"Escapes curly braces before calling FormattableStringFactory.Create when interpolated string literal is typed as FormattableString"
 3551,buildDuplicateFile,"The source file '%s' (at position %d/%d) already appeared in the compilation list (at position %d/%d). Please verify that it is included only once in the project file."
-featureEscapeBracesInFormattableString,"Escapes curly braces before calling FormattableStringFactory.Create when interpolated string literal is typed as FormattableString"
\ No newline at end of file
+3552,chkConstructorWithArgumentsOnStaticClasses,"If a type uses both [<Sealed>] and [<AbstractClass>] attributes, it means it is static. constructor with arguments is not allowed."
+3553,chkAdditionalConstructorOnStaticClasses,"If a type uses both [<Sealed>] and [<AbstractClass>] attributes, it means it is static. additional constructor is not allowed."
```
Comment (0101) [inline src/Compiler/FSComp.txt:1670]:
```suggestion
3552,chkConstructorWithArgumentsOnStaticClasses,"If a type uses both [<Sealed>] and [<AbstractClass>] attributes, it means it is static. Constructor with arguments is not allowed."
3553,chkAdditionalConstructorOnStaticClasses,"If a type uses both [<Sealed>] and [<AbstractClass>] attributes, it means it is static. Additional constructor is not allowed."
```
Capitalize start of sentence.

**Review by 0101 — CHANGES_REQUESTED** [review]
Looks good. But I think we should fix the capitalization in the error messages.

**Review by 0101 — APPROVED** [review]

**Review by T-Gro — APPROVED** [review]

**Pull request merged**